examples/cpp/helloworld/greeter_async_bidi_server.cc

```diff
@@ -0,0 +1,197 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+
+#include <grpc++/grpc++.h>
+
+#include "hellostreamingworld.grpc.pb.h"
+
+using grpc::Server;
+using grpc::ServerAsyncReaderWriter;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::ServerCompletionQueue;
+using grpc::Status;
+using hellostreamingworld::HelloRequest;
+using hellostreamingworld::HelloReply;
+using hellostreamingworld::MultiGreeter;
+
+enum class Type { READ = 1, WRITE = 2, CONNECT = 3, DONE = 4, FINISH = 5 };
+
+// NOTE: This is a complex example for an asynchronous, bidirectional streaming
+// server. For a simpler example, start with the
+// greeter_server/greeter_async_server first.
+
+// Most of the logic is similar to AsyncBidiGreeterClient, so follow that class
+// for detailed comments. Two main differences between the server and the client
+// are: (a) Server cannot initiate a connection, so it first waits for a
+// 'connection'. (b) Server can handle multiple streams at the same time, so
+// the completion queue/server have a longer lifetime than the client(s).
+class AsyncBidiGreeterServer {
+ public:
+  AsyncBidiGreeterServer() {
+    // In general avoid setting up the server in the main thread (specifically,
+    // in a constructor-like function such as this). We ignore this in the
+    // context of an example.
+    std::string server_address("0.0.0.0:50051");
+
+    ServerBuilder builder;
+    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+    builder.RegisterService(&service_);
+    cq_ = builder.AddCompletionQueue();
+    server_ = builder.BuildAndStart();
+
+    // This initiates a single stream for a single client. To allow multiple
+    // clients in different threads to connect, simply 'request' from the
+    // different threads. Each stream is independent but can use the same
+    // completion queue/context objects.
+    stream_.reset(
+        new ServerAsyncReaderWriter<HelloReply, HelloRequest>(&context_));
```

Comment:
So I would say that this is fundamentally different from (and less useful than) the sync server example because it only allows 1 RPC at a time to be outstanding. If you only have 1 RPC outstanding, you should definitely use sync APIs because they will make your life much easier. C++ async API is only attractive when you have more than a few hundred (or more than a few thousand, in more likelihood) concurrent RPCs